pkg/admission/server.go

```diff
@@ -0,0 +1,142 @@
+package admission
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	admission "k8s.io/api/admission/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"net/http"
+	"path/filepath"
+	sav1 "sigs.k8s.io/service-apis/apis/v1alpha1"
+	"k8s.io/klog/v2"
+)
+
+const (
+	tlsDir      = `/run/secrets/tls`
+	tlsCertFile = `tls.crt`
+	tlsKeyFile  = `tls.key`
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+
+// ServeHTTP parses AdmissionReview requests and responds back
+// with the validation result of the entity.
+func ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Body == nil {
+		// log Error("received request with empty body")
+		http.Error(w, "admission review object is missing",
+			http.StatusBadRequest)
+		return
+	}
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		klog.Errorf("failed to read request from client: %v\n", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	review := admission.AdmissionReview{}
+	if err := json.Unmarshal(data, &review); err != nil {
+		klog.Errorf("failed to parse AdmissionReview object: %v\n", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	response, err := handleValidation(*review.Request)
+	if err != nil {
+		klog.Errorf("failed to run validation: %v\n", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	review.Response = response
+	data, err = json.Marshal(review)
+	if err != nil {
+		klog.Errorf("failed to marshal response: %v\n", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	_, err = w.Write(data)
+	if err != nil {
+		klog.Errorf("failed to write response: %v\n", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+
+var (
+	httpRoute = meta.GroupVersionResource{
+		Group:    sav1.SchemeGroupVersion.Group,
+		Version:  sav1.SchemeGroupVersion.Version,
+		Resource: "httproutes",
+	}
+	gateway = meta.GroupVersionResource{
+		Group:    sav1.SchemeGroupVersion.Group,
+		Version:  sav1.SchemeGroupVersion.Version,
+		Resource: "gateway",
+	}
+)
+
+func handleValidation(request admission.AdmissionRequest) (
+	*admission.AdmissionResponse, error) {
+	var response admission.AdmissionResponse
+	var ok bool
+	var message string
+	var err error
+
+	if request.Resource == httpRoute {
+		hRoute := sav1.HTTPRoute{}
+		deserializer := codecs.UniversalDeserializer()
+		_, _, err = deserializer.Decode(request.Object.Raw, nil, &hRoute)
+		if err != nil {
+			return nil, err
+		}
+		// TODO(cmluciano) Add update operations
+		switch request.Operation {
+		case admission.Create:
+			ok, message, err = ValidateHTTPRoute(hRoute)
```

Comment:
Would we need to do anything different for updates?

Reply:
Yes I need to add the update part as well